Kubernetes guide for Weave 1.1

errordeveloper opened the pull request:
There will more specific statements we would like to add to introduction and summary, but here is the starting point.


## Introduction

With Weave 1.1 release users can utilise Docker API proxy and IP address allocator with Kubernetes, thereby removing the need to configure Docker daemon in a certain way, e.g. with statically allocated subnets for each host etc.

[This is a reworked version of our old Kubernetes guide](http://blog.weave.works/2014/11/11/weave-for-kubernetes/). Many new features had been added to Weave since the time of writing, namely Weave as well as Kubernetes have both become much more mature and gone past 1.0 releases.

## Conclusion

Say that Weave is the only networking solution for Kubernetes that offers a very easy to use encryption. Weave can potentially allow users to access un-orchestrated and even uncontainerized services (such as legacy databases), as well as services on running on other orchestrators, such as Mesos.

## Instructions

```
> git clone https://github.com/weaveworks/guides
> git checkout kubernetes
> cd kubernetes/coreos
> vagrant up
> vagrant ssh kube-00
kube-01 > journalctl -f -u install-kubernetes       ## wait for kubernetes binaries to get installed
kube-01 > kubectl create -f /etc/kubernetes/addons  ## deploy DNS addon (when status column goes ready)
kube-01 > kubectl get pods -w --all-namespaces      ## wait for DNS pods
kube-01 > kubectl create -f guestbook-example       ## deploy the example Guestbook PHP
kube-01 > kubectl get pods -w                       ## wait for all app pods to get deployed
```

Next you should be able to hit the app on `172.17.8.101` with port number that was printed when you ran `kubectl create -f guestbook-example`.

Other useful commands:
  - `etcdctl cluster-health`
  - `systemctl status kube-*`
  - `kubectl get nodes`

rade commented:
Is there a way to setup weave+k8s that doesn't involve us having copies of 400+ lines of config, the vast majority of which has nothing to do with weave?

rade commented on kubernetes/coreos/kubernetes-cluster.yaml:
```diff
@@ -0,0 +1,369 @@
+        Requires=install-weave.service
+        [Service]
+        EnvironmentFile=-/etc/weave.env
+        ExecStartPre=/opt/bin/weave launch-proxy --rewrite-inspect --without-dns
```
why is there no `weave setup` here, as there is for `weave.service`?

errordeveloper replied:
That one used to be necessary, can be changed now.

On Sun, 6 Sep 2015 08:43 Matthias Radestock notifications@github.com
wrote:

> In kubernetes/coreos/kubernetes-cluster.yaml
> https://github.com/weaveworks/guides/pull/52#discussion_r38818579:
> 
> > -        WantedBy=weave-network.target
> > -        WantedBy=weave.service
> >   +
> > -    - name: weaveproxy.service
> > -      enable: true
> > -      content: |
> > -        [Unit]
> > -        After=install-weave.service
> > -        After=docker.service
> > -        Description=Weave Network Router
> > -        Documentation=http://docs.weave.works/
> > -        Requires=docker.service
> > -        Requires=install-weave.service
> > -        [Service]
> > -        EnvironmentFile=-/etc/weave.env
> > -        ExecStartPre=/opt/bin/weave launch-proxy --rewrite-inspect --without-dns
> 
> why is there no weave setup here, as there is for weave.service?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/weaveworks/guides/pull/52/files#r38818579.

rade commented on kubernetes/coreos/kubernetes-cluster.yaml:
```diff
@@ -0,0 +1,369 @@
+        TimeoutStartSec=0
+        EnvironmentFile=-/etc/weave.env
+        EnvironmentFile=/etc/weave.%H.env
+        ExecStartPre=/opt/bin/weave setup
```
Wouldn't it make sense to invoke `setup` in `install-weave.service` instead?

errordeveloper replied:
Done.

rade commented on kubernetes/coreos/kubernetes-cluster.yaml:
```diff
@@ -0,0 +1,369 @@
+        ExecStartPre=/usr/bin/curl \
+          --silent \
+          --location \
+          https://raw.github.com/weaveworks/weave/master/weave \
```
shouldn't this fetch from `git.io/weave-snapshot` instead?

errordeveloper replied:
Yes, I just forgot what this URL was. Fixed now.

rade commented on kubernetes/coreos/kubernetes-cluster.yaml:
```diff
@@ -0,0 +1,369 @@
+        [Unit]
+        After=install-weave.service
+        After=docker.service
+        Description=Weave Network Router
```
That description is the same as for `weave.service`, which is confusing.

errordeveloper commented:
That's a good point, as we do not depend on Docker's config, we can care
less about however daemon is configured. The question is where exactly
should we take it? One approach would be to use Atomic, which comes with
k8s binaries and show users how flannel can be replaced with Weave. Another
way would involve running k8s components in containers on Weave, which is
one I was gonna do next already anyway. And one more approach that would
make sense to new users, it would involve a bold assumption regarding how
things are configured and which distro is used, it would basically mean the
user has to do the automation, if they wish to. There are potentially some
kind of approaches that would be in between of the above, I'll make few
considerations and post those later.

On Sun, 6 Sep 2015 08:41 Matthias Radestock <notifications@github.com>
wrote:

> Is there a way to setup weave+k8s that doesn't involve us having copies of
> 400+ lines of config, the vast majority of which has nothing to do with
> weave?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/weaveworks/guides/pull/52#issuecomment-138050560>.
>

rade commented:
> The question is where exactly should we take it?

I was kinda hoping that the steps could be:

1. install k8s using their instructions
2. install weave using our instructions here

errordeveloper commented:
> 1. install k8s using their instructions

There is a wild variety of options, hence I have forked one of the options earlier on. I can take a look at where things stand now, although I don't think there had been much of improvement.

rade commented:
> There is a wild variety of options

Yes, there is.

> hence I have forked one of the options earlier on.

By picking one we automatically exclude the others, which is not good.

abuehrle commented:
@errordeveloper have a look.  I talk about encryption but we don't enable it in the example. Should we change that?

paulbellamy commented on kubernetes/coreos/README.md:
```diff
@@ -0,0 +1,199 @@
+~~~bash
+> vagrant ssh kube-01
+
+kube-01 > weave setup
```
Don't we mean `weave launch --something --something` ? Or is `install-kubernetes` later secretly doing some of this?

abuehrle commented:
@paulbellamy  not sure.  Should it just be `weave launch`, then, since Vagrant already sets up weave?

errordeveloper commented:
It's all done, there are no `weave` commands to call, other then `weave
status`. If we were to launch anything manually, it would make things to
complicated, as there 3 VMs to handle. We can do a single VM manual guide,
later maybe.

On Tue, 8 Sep 2015 13:06 Anita Buehrle <notifications@github.com> wrote:

> @paulbellamy <https://github.com/paulbellamy> not sure. Should it just be weave
> launch, then, since Vagrant already sets up weave?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/weaveworks/guides/pull/52#issuecomment-138519192>.
>

abuehrle commented:
Thanks Ilya.  I have changed it to say ‘weave status’.  If there are no more comments, I think this is ready to be merged. 


> On Sep 8, 2015, at 7:44 AM, Ilya Dmitrichenko <notifications@github.com> wrote:
> 
> It's all done, there are no `weave` commands to call, other then `weave
> status`. If we were to launch anything manually, it would make things to
> complicated, as there 3 VMs to handle. We can do a single VM manual guide,
> later maybe.
> 
> On Tue, 8 Sep 2015 13:06 Anita Buehrle <notifications@github.com> wrote:
> 
> > @paulbellamy <https://github.com/paulbellamy> not sure. Should it just be weave
> > launch, then, since Vagrant already sets up weave?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/weaveworks/guides/pull/52#issuecomment-138519192>.
> >
> —
> Reply to this email directly or view it on GitHub <https://github.com/weaveworks/guides/pull/52#issuecomment-138548730>.
>

errordeveloper commented:
I think I have now addresses all of the comments to do with `kubernetes-cluster.yaml`. The line comments  don't seem to hide themselves, probably due to my comments being posted via email and github not handling that as easily.

errordeveloper commented:
> Is there a way to setup weave+k8s that doesn't involve us having copies of 400+ lines of config, the vast majority of which has nothing to do with weave?

I think it's best to merge this as it is and address #63 next.
